### PR TITLE
Gen 5 ID Generator should ignore XOR

### PR DESCRIPTION
**Pull request by ShinySylveon04**
This PR is for issue #190 to only show SIDs that would make the given PID shiny without the xor check. Please let me know if there's anything that needs changing.

Thanks to @Real96 for the help with this!

### CODE REVIEW

File: Source/Forms/Gen5/IDs5.ui [hunk 1]
```diff
@@ -426,6 +426,13 @@
         </property>
        </widget>
       </item>
+      <item row="5" column="0">
+       <widget class="QCheckBox" name="checkBoxXOR">
+        <property name="text">
+         <string>Show XOR PID Results</string>
```
Comment (Admiral-Fish) [inline Source/Forms/Gen5/IDs5.ui:432]: This will probably be very confusing to users that are new to Gen 5. It would probably be better to give some indication of what they want to rng (stationary/wild vs egg/event/) or whatever the combos are

File: Source/Forms/Gen5/IDs5.cpp [hunk 2]
```diff
@@ -158,20 +159,24 @@ void IDs5::search()
...
+    connect(timer, &QTimer::destroyed,
+            [=]
+            {
+                ui->pushButtonSearch->setEnabled(true);
+                ui->pushButtonFind->setEnabled(true);
+                ui->pushButtonCancel->setEnabled(false);
+                model->addItems(searcher->getResults());
+                ui->progressBar->setValue(searcher->getProgress());
+                delete searcher;
+            });
```
Comment (Admiral-Fish) [inline Source/Forms/Gen5/IDs5.cpp:179]: Please use the .clang-format file in the repo. When I save it does not format to this

**Review by Admiral-Fish — CHANGES_REQUESTED** [review]

File: Source/Forms/Gen5/IDs5.cpp (hunk 2 above)
Comment (Admiral-Fish) [reply Source/Forms/Gen5/IDs5.cpp:186]: LambdaBodyIndentation: OuterScope

File: Source/Forms/Gen5/IDs5.hpp [hunk 3]
```diff
@@ -52,6 +52,7 @@ class IDs5 : public QWidget
     void setupModels();
 
 private slots:
+    void setXOR(bool);
```
Comment (Admiral-Fish) [inline Source/Forms/Gen5/IDs5.hpp:55]: Please include the variable name

File: Source/Forms/Gen5/IDs5.cpp [hunk 4]
```diff
@@ -115,6 +116,12 @@ void IDs5::setupModels()
     setting.endGroup();
 }
 
+void IDs5::setXOR(bool checked)
+{
+    ui->checkBoxXOR->setChecked(false);
```
Comment (Admiral-Fish) [inline Source/Forms/Gen5/IDs5.cpp:121]: I don't think this line is necessary

File: Source/Forms/Gen5/IDs5.cpp [hunk 5]
```diff
@@ -115,6 +116,12 @@ void IDs5::setupModels()
...
+    ui->checkBoxXOR->setEnabled(checked ? ui->checkBoxPID->isChecked() : false);
```
Comment (Admiral-Fish) [inline Source/Forms/Gen5/IDs5.cpp:122]: Shouldn't `checked` already contain the value of `ui->checkBoxPID->isChecked()`

File: Source/Core/Gen5/Generators/IDGenerator5.cpp [hunk 6]
```diff
@@ -50,19 +49,13 @@ std::vector<IDState5> IDGenerator5::generate(u64 seed, u32 pid, bool checkPID)
...
-                u16 actualPSV = (idBit ^ pidBit) ? xorPSV : psv;
-                if ((actualPSV >> 3) == state.getTSV())
-                {
-                    state.setSeed(seed);
-                    states.emplace_back(state);
-                }
-            }
-            else
+            // Check if PID will be modified by TID/SID combo if XOR box is checked
+            bool isShinyXor = checkXOR && !(idBit ^ pidBit) && (psv >> 3) == state.getTSV();
```
Comment (Admiral-Fish) [inline Source/Core/Gen5/Generators/IDGenerator5.cpp:55]: What is `!(idBit ^ pidBit)` checking for. Please change it to what was described here https://github.com/Admiral-Fish/PokeFinder/pull/191#commitcomment-57075211

**Review by Admiral-Fish — CHANGES_REQUESTED** [review]

File: Source/Core/Gen5/Generators/IDGenerator5.cpp (hunk 6 above)
Comment (Real96) [reply Source/Core/Gen5/Generators/IDGenerator5.cpp:55]: > What is `!(idBit ^ pidBit)` checking for. Please change it to what was described here [#191 (comment)](https://github.com/Admiral-Fish/PokeFinder/pull/191#commitcomment-57075211)

It avoids showing IDs combo that will make the PID ^ 80000000

File: Source/Forms/Gen5/IDs5.cpp (hunk 4 above)
Comment (ShinySylveon04) [reply Source/Forms/Gen5/IDs5.cpp:121]: I added it to prevent this from happening: 
image

If both PID and XOR are checked, then PID is unchecked, the XOR checkbox will still be checked. Is there another way to prevent this from happening?

File: Source/Core/Gen5/Generators/IDGenerator5.hpp [hunk 7]
```diff
@@ -28,7 +28,7 @@ class IDGenerator5 : public IDGenerator
 public:
     IDGenerator5() = default;
     IDGenerator5(u32 initialAdvances, u32 maxAdvances, const IDFilter &filter);
-    std::vector<IDState5> generate(u64 seed, u32 pid = 0, bool checkPID = false);
+    std::vector<IDState5> generate(u64 seed, u32 pid = 0, bool checkPID = false, bool checkXOR = false);
```
Comment (Admiral-Fish) [inline Source/Core/Gen5/Generators/IDGenerator5.hpp:31]: I know part of this is me, but neither of these need a default value. Please remove it from both checkPID and checkXOR.

File: Source/Core/Gen5/Generators/IDGenerator5.cpp [hunk 8]
```diff
@@ -33,12 +33,11 @@ std::vector<IDState5> IDGenerator5::generate(u64 seed, u32 pid, bool checkPID)
 
     bool pidBit = (pid >> 31) ^ (pid & 1);
     u16 psv = (pid >> 16) ^ (pid & 0xffff);
-    u16 xorPSV = psv ^ 0x8000;
```
Comment (Admiral-Fish) [inline Source/Core/Gen5/Generators/IDGenerator5.cpp:36]: Why was this removed

File: Source/Core/Gen5/Generators/IDGenerator5.cpp [hunk 9]
```diff
@@ -50,19 +49,13 @@ std::vector<IDState5> IDGenerator5::generate(u64 seed, u32 pid, bool checkPID)
...
-                u16 actualPSV = (idBit ^ pidBit) ? xorPSV : psv;
-                if ((actualPSV >> 3) == state.getTSV())
-                {
-                    state.setSeed(seed);
-                    states.emplace_back(state);
-                }
-            }
-            else
+            // Check if PID will be modified by TID/SID combo if XOR box is checked
+            bool isShinyXor = checkXOR && (idBit == pidBit) && (psv >> 3) == state.getTSV();
```
Comment (Admiral-Fish) [inline Source/Core/Gen5/Generators/IDGenerator5.cpp:55]: I don't think this check is correct. See above comment about removing `xorPSV`. You need to check which PSV you are using depending on the result of (idBit == pidBit).

I think something like this would be correct for this block. Very verbose but it should give the right idea.
```
if (filter.compare(state))
{
    if (checkPID)
    {
        if (checkXOR)
        {
            bool idBit = (tid & 1) ^ (sid & 1);
            u16 actualPSV = (idBit ^ pidBit) ? xorPSV : psv;
            if ((actualPSV >> 3) == state.getTSV())
            {
                state.setSeed(seed);
                states.emplace_back(state);
            }
        }
        else
        {
            if ((psv >> 3) == state.getTSV())
            {
                state.setSeed(seed);
                states.emplace_back(state);
            }
        }
    }
    else
    {
        state.setSeed(seed);
        states.emplace_back(state);
    }
}
```

**Review by Admiral-Fish — CHANGES_REQUESTED** [review]

File: Source/Core/Gen5/Generators/IDGenerator5.cpp (hunk 9 above)
Comment (zaksabeast) [reply Source/Core/Gen5/Generators/IDGenerator5.cpp:55]: What are thoughts on simplifying it to remove nesting and make the code a bit more DRY:
```cpp
if (filter.compare(state)) {
    bool idBit = (tid & 1) ^ (sid & 1);

    // Check if PID will be modified by TID/SID combo if XOR box is checked
    u16 psvToTest = !checkXOR || (checkXOR && idBit == pidBit) ? psv : xorPSV;
    bool isShiny = (psvToTest >> 3) == state.getTSV();

    if (!checkPID || isShiny)
    {
        state.setSeed(seed);
        states.emplace_back(state);
    }
}
```

File: Source/Core/Gen5/Generators/IDGenerator5.cpp (hunk 8 above)
Comment (Real96) [reply Source/Core/Gen5/Generators/IDGenerator5.cpp:36]: Cause we don't need to check IDs combo for the xored PID. We're trying to get IDs combo to make the inserted PID shiny, not the xored PID

File: Source/Core/Gen5/Generators/IDGenerator5.cpp (hunk 9 above)
Comment (Real96) [reply Source/Core/Gen5/Generators/IDGenerator5.cpp:55]: > I don't think this check is correct. See above comment about removing `xorPSV`. You need to check which PSV you are using depending on the result of (idBit == pidBit).
> 
> I think something like this would be correct for this block. Very verbose but it should give the right idea.
> 
> ```
> if (filter.compare(state))
> {
>     if (checkPID)
>     {
>         if (checkXOR)
>         {
>             bool idBit = (tid & 1) ^ (sid & 1);
>             u16 actualPSV = (idBit ^ pidBit) ? xorPSV : psv;
>             if ((actualPSV >> 3) == state.getTSV())
>             {
>                 state.setSeed(seed);
>                 states.emplace_back(state);
>             }
>         }
>         else
>         {
>             if ((psv >> 3) == state.getTSV())
>             {
>                 state.setSeed(seed);
>                 states.emplace_back(state);
>             }
>         }
>     }
>     else
>     {
>         state.setSeed(seed);
>         states.emplace_back(state);
>     }
> }
> ```

We don't need this check. We need IDs combo for our wanted PID, not for the xored one.
If you leave the check, you'll get IDs combo to make the wanted PID shiny, but you'll get also IDs combot that will make the PID^0x80000 to be shiny. We don't want the last thing

Comment (Admiral-Fish) [reply Source/Core/Gen5/Generators/IDGenerator5.cpp:55]: Yeah looks good zak

Comment (Parzival5) [reply Source/Core/Gen5/Generators/IDGenerator5.cpp:55]: If we check (psv ^ 0x8000), the user will always be able to get a shiny but in the half of the cases he will not get the specified pid.

Maybe something like this would be fine:
```c++
if (filter.compare(state)) {
    
    bool isShiny = (psv >> 3) == state.getTSV();

    // Check if PID is possible with TID/SID combo if Wild/Stationary box is checked
    if (checkXOR)
        isShiny &= ((tid & 1) ^ (sid & 1)) == pidBit;
    
    if (!checkPID || isShiny)
    {
        state.setSeed(seed);
        states.emplace_back(state);
    }
}
````

Comment (zaksabeast) [reply Source/Core/Gen5/Generators/IDGenerator5.cpp:55]: We should be able to replace the `&=` with `=` to be a bit more readable:
```cpp
isShiny = ((tid & 1) ^ (sid & 1)) == pidBit;
```

Due to the if statement, the left hand side of that calculation will always be `true`.

File: Source/Forms/Gen5/IDs5.cpp (hunk 2 above)
Comment (ShinySylveon04) [reply Source/Forms/Gen5/IDs5.cpp]: This should be formatted correctly now

File: Source/Core/Gen5/Generators/IDGenerator5.hpp (hunk 7 above)
Comment (ShinySylveon04) [reply Source/Core/Gen5/Generators/IDGenerator5.hpp]: Default values have been removed

File: Source/Core/Gen5/Generators/IDGenerator5.cpp (hunk 9 above)
Comment (ShinySylveon04) [reply Source/Core/Gen5/Generators/IDGenerator5.cpp]: The above change has been done

File: Source/Core/Gen5/Generators/IDGenerator5.hpp [hunk 10]
```diff
@@ -23,12 +23,11 @@
...
 #include <Core/Parents/Generators/IDGenerator.hpp>
 
-class IDGenerator5 : public IDGenerator
-{
+class IDGenerator5 : public IDGenerator {
 public:
     IDGenerator5() = default;
     IDGenerator5(u32 initialAdvances, u32 maxAdvances, const IDFilter &filter);
-    std::vector<IDState5> generate(u64 seed, u32 pid = 0, bool checkPID = false);
+    std::vector<IDState5> generate(u64 seed, bool checkPID, bool checkXOR, u32 pid = 0);
```
Comment (Admiral-Fish) [inline Source/Core/Gen5/Generators/IDGenerator5.hpp:30]: See comment in IDs5.cpp, I remember why these had defaults

File: Source/Forms/Gen5/IDs5.cpp [hunk 11]
```diff
@@ -213,7 +224,7 @@ void IDs5::find()
             u64 seed = sha.hashSeed();
 
             generator.setInitialAdvances(flag ? Utilities::initialAdvancesBWID(seed) : Utilities::initialAdvancesBW2ID(seed));
-            auto states = generator.generate(seed);
+            auto states = generator.generate(seed, usePID, useXOR, pid);
```
Comment (Admiral-Fish) [inline Source/Forms/Gen5/IDs5.cpp:227]: Okay I now remember why it had default parameters. Could you put it back that way it was, my bad.

File: Source/Forms/Gen5/IDs5.cpp [hunk 12]
```diff
@@ -188,6 +196,9 @@ void IDs5::find()
     u8 minSecond = ui->spinBoxMinSecond->value();
     u8 maxSecond = ui->spinBoxMaxSecond->value();
     u32 maxAdvance = ui->textBoxSeedFinderMaxAdvances->getUInt();
+    u32 pid = ui->textBoxPID->getUInt();
+    bool usePID = ui->checkBoxPID->isChecked();
+    bool useXOR = ui->checkBoxXOR->isChecked();
```
Comment (Admiral-Fish) [inline Source/Forms/Gen5/IDs5.cpp:201]: These can be removed once defaults are back

File: Source/Core/Gen5/Searchers/IDSearcher5.hpp [hunk 13]
```diff
@@ -26,11 +26,10 @@
 #include <atomic>
 #include <mutex>
 
-class IDSearcher5
-{
+class IDSearcher5 {
```
Comment (Admiral-Fish) [inline Source/Core/Gen5/Searchers/IDSearcher5.hpp:29]: Put bracket on a new line

File: Source/Core/Gen5/Generators/IDGenerator5.hpp [hunk 14]
```diff
@@ -23,12 +23,11 @@
 #include <Core/Gen5/States/IDState5.hpp>
 #include <Core/Parents/Generators/IDGenerator.hpp>
 
-class IDGenerator5 : public IDGenerator
-{
+class IDGenerator5 : public IDGenerator {
```
Comment (Admiral-Fish) [inline Source/Core/Gen5/Generators/IDGenerator5.hpp:26]: Put bracket on newline

File: Source/Core/Gen5/Generators/IDGenerator5.cpp [hunk 15]
```diff
@@ -50,19 +49,13 @@ std::vector<IDState5> IDGenerator5::generate(u64 seed, u32 pid, bool checkPID)
...
-                if ((actualPSV >> 3) == state.getTSV())
-                {
-                    state.setSeed(seed);
-                    states.emplace_back(state);
-                }
-            }
-            else
+            // Check if PID is possible with TID/SID combo if Wild/Stationary box is checked
+            if (isShiny && checkXOR) // We need to do the check only if it was shiny first
+                isShiny = ((tid & 1) ^ (sid & 1)) == pidBit;
```
Comment (Admiral-Fish) [inline Source/Core/Gen5/Generators/IDGenerator5.cpp:56]: I literally have no clue what this is doing. Why was it changed from what zak and I agreed on

Also make sure if statements have opening and closing brackets

File: .clang-format [hunk 16]
```diff
@@ -4,7 +4,13 @@ BasedOnStyle: WebKit
...
 AllowShortFunctionsOnASingleLine: None
 AlwaysBreakTemplateDeclarations: Yes
-BreakBeforeBraces: Allman
+BreakBeforeBraces: Custom
+BraceWrapping:
+  BeforeElse: true
+  BeforeLambdaBody: false
+  AfterCaseLabel: true
+  AfterControlStatement: true
+  AfterFunction: true
```
Comment (Admiral-Fish) [inline .clang-format:13]: Changing to what I have below should help fix the formatting issues I mentioned

```
BreakBeforeBraces: Custom
BraceWrapping:
  AfterCaseLabel: true
  AfterClass: true
  AfterControlStatement: true
  AfterEnum: true
  AfterFunction: true
  AfterNamespace: true
  AfterStruct: true
  BeforeElse: true
  BeforeLambdaBody: false
```

**Review by Admiral-Fish — CHANGES_REQUESTED** [review]
